stan/math/prim/fun/eval.hpp

```diff
@@ -0,0 +1,39 @@
+#ifndef STAN_MATH_PRIM_FUN_EVAL_HPP
+#define STAN_MATH_PRIM_FUN_EVAL_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <cstddef>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Inputs that are not Eigen types are forwarded unmodified
+ *
+ * @tparam T Input type
+ * @param[in] arg Input argument
+ * @return Forwarded input argument
+ **/
+template <typename T, require_not_eigen_t<T>* = nullptr>
+inline decltype(auto) eval(T&& arg) {
+  return std::forward<T>(arg);
+}
+
+/**
+ * Inputs that are Eigen types are eval'd and returned
+ *
+ * @tparam T Input type
+ * @param[in] arg Input argument
+ * @return Eval'd argument
+ **/
+template <typename T, require_eigen_t<T>* = nullptr>
+inline decltype(auto) eval(T arg) {
+  return arg.eval();
```

Comment:
I had a feeling we'd need that sooner or later. 

Why pass by value? I think const ref should be used here.

Also I think the first overload should be used for plain types (you can use `require_same<std::decay_t<T>, plain_type_t<T>>` instead of `require_eigen`), as their eval just returns a const refrence and original object might go out of scope before the reference is used for rvalue arguments.

Reply:
```require_same<std::decay_t<T>, plain_type_t<T>>```

That's a better way to implement this. Now not all eigen types get `const` added out front.

```Why pass by value?```

Wasn't thinking right. const ref is right here.